common/k8s/src/metrics_stats_aggregator.rs

```diff
@@ -0,0 +1,672 @@
+};
+use serde_json::Value;
+
+use std::collections::HashMap;
+use std::time::Duration;
+use tokio::time::sleep;
+
+use crate::kube_stats::{
+    cluster_stats::ClusterStats,
+    container_stats::ContainerStats,
+    controller_stats::ControllerStats,
+    extended_pod_stats::ExtendedPodStats,
+    node_stats::{NodeContainerStats, NodePodStats, NodeStats},
+    pod_stats::PodStats,
+};
+
+pub struct MetricsStatsAggregator {
+    pub client: Client,
+}
+
+impl MetricsStatsAggregator {
+    pub fn new(client: Client) -> Self {
+        Self { client }
+    }
+
+    pub fn start_metrics_call_task(self) -> impl Stream<Item = LineBuilder> {
+        stream::unfold(self.client, |client| async {
+            sleep(Duration::from_millis(30000)).await;
+            Some((
+                match self::process_reporter_info(client.clone()).await {
+                    Ok((pods_strings, node_strings, cluster_stats_string)) => Some(
+                        stream::iter(
+                            pods_strings
+                                .into_iter()
+                                .chain(node_strings.into_iter())
+                                .chain(Some(cluster_stats_string).into_iter()),
+                        )
+                        .map(|line| {
+                            LineBuilder::new()
+                                .line(line)
+                                .file("logdna-reporter".to_string())
+                        }),
+                    ),
+                    Err(e) => {
+                        error!("Failed To Gather Metrics Server Info {}", e);
+                        None
+                    }
+                },
+                client,
+            ))
+        })
+        .filter_map(|x| async { x })
+        .flatten()
+    }
+}
+
+async fn process_reporter_info(
+    client: Client,
+) -> anyhow::Result<(Vec<String>, Vec<String>, String)> {
+    let pods = self::get_all_pods(client.clone()).await?;
+    let nodes = self::get_all_nodes(client.clone()).await?;
+    let pod_metrics = self::call_metric_api("PodMetrics", client.clone()).await?;
+    let node_metrics = self::call_metric_api("NodeMetrics", client.clone()).await?;
+
+    let mut controller_map: HashMap<String, ControllerStats> = HashMap::new();
+    let mut node_pod_counts_map: HashMap<String, NodePodStats> = HashMap::new();
+    let mut node_container_counts_map: HashMap<String, NodeContainerStats> = HashMap::new();
+    let mut pod_usage_map: HashMap<String, Value> = HashMap::new();
+    let mut node_usage_map: HashMap<String, Value> = HashMap::new();
+
+    let mut extended_pod_stats: Vec<ExtendedPodStats> = Vec::new();
+    let mut node_stats: Vec<NodeStats> = Vec::new();
+
+    build_pod_metric_map(pod_metrics, &mut pod_usage_map);
+    process_pods(
+        pods,
+        &mut controller_map,
+        pod_usage_map,
+        &mut extended_pod_stats,
+        &mut node_pod_counts_map,
+        &mut node_container_counts_map,
+    );
```

Comment:
always successful?

Reply:
It can fail - but the error will propagate to the calling function - which handles. 

This being a private function - it won't be called elsewhere.

Reply:
misunderstood this comment - moved to private chat